src/dtx/dtx_common.c

```diff
@@ -0,0 +1,528 @@
+				dte, rc, cont_hdl->sch_pool->spc_map_version);
+		D_FREE(dte);
+	} while (rc >= 0);
+
+	if (rc < 0)
+		D_ERROR(DF_UUID": Fail to flush CoS cache: rc = %d\n",
+			DP_UUID(cont->sc_uuid), rc);
+
+	cont->sc_dtx_flushed = 1;
+}
+
+void
+dtx_batched_commit(void *arg)
+{
+	struct dss_module_info	*dmi = dss_get_module_info();
+
+	while (1) {
+		ABT_bool			 state;
+		struct dtx_batched_commit_args	*dbca;
+		struct ds_cont_hdl		*cont_hdl;
+		struct ds_cont			*cont;
+		struct dtx_entry		*dte = NULL;
+		struct dtx_stat			 stat = { 0 };
+		int				 rc;
+
+		if (d_list_empty(&dmi->dmi_dtx_batched_list))
+			goto check;
+
+		dbca = d_list_entry(dmi->dmi_dtx_batched_list.next,
+				    struct dtx_batched_commit_args, dbca_link);
+		cont_hdl = dbca->dbca_cont_hdl;
+		cont = cont_hdl->sch_cont;
+		if (cont->sc_closing) {
+			dtx_flush_committable(cont_hdl);
+			/* XXX: There may be resent-close during DTX flush,
+			 *	so NOT remove the @dbca from the list until
+			 *	DTX flushed.
+			 */
+			d_list_del(&dbca->dbca_link);
+			/* drop the ref for dtx_batched_commit_register(). */
+			ds_cont_hdl_put(cont_hdl);
+			D_FREE_PTR(dbca);
+			goto check;
+		}
+
+		d_list_move_tail(&dbca->dbca_link, &dmi->dmi_dtx_batched_list);
+		vos_dtx_stat(cont->sc_hdl, &stat);
+
+		if ((stat.dtx_committable_count > DTX_THRESHOLD_COUNT) ||
+		    (stat.dtx_oldest_committable_time != 0 &&
+		     dtx_hlc_age2sec(stat.dtx_oldest_committable_time) >
+		     DTX_COMMIT_THRESHOLD_AGE)) {
+			rc = vos_dtx_list_committable(cont->sc_hdl,
+						DTX_THRESHOLD_COUNT, &dte);
+			if (rc > 0) {
+				rc = dtx_commit(cont_hdl->sch_pool->spc_uuid,
+					cont->sc_uuid, dte, rc,
+					cont_hdl->sch_pool->spc_map_version);
+				D_FREE(dte);
+
+				if (cont->sc_closing) {
+					dtx_flush_committable(cont_hdl);
+					d_list_del(&dbca->dbca_link);
+					ds_cont_hdl_put(cont_hdl);
+					D_FREE_PTR(dbca);
+					goto check;
+				}
+
+				if (!cont->sc_dtx_aggregating)
+					vos_dtx_stat(cont->sc_hdl, &stat);
+			}
+		}
+
+		if (!cont->sc_dtx_aggregating &&
+		    ((stat.dtx_committed_count > DTX_AGG_THRESHOLD_CNT) ||
+		     (stat.dtx_oldest_committed_time != 0 &&
+		      dtx_hlc_age2sec(stat.dtx_oldest_committed_time) >
+				DTX_AGG_THRESHOLD_AGE_UPPER))) {
+			ds_cont_get(cont);
+			cont->sc_dtx_aggregating = 1;
+			rc = dss_ult_create(dtx_aggregate, cont, DSS_ULT_SELF,
+					    dmi->dmi_tgt_id, 0, NULL);
+			if (rc != 0) {
+				cont->sc_dtx_aggregating = 0;
+				ds_cont_put(cont);
+			}
+		}
+
+check:
+		rc = ABT_future_test(dmi->dmi_xstream->dx_shutdown, &state);
+		D_ASSERTF(rc == ABT_SUCCESS, "%d\n", rc);
+		if (state == ABT_TRUE) {
+			while (!d_list_empty(&dmi->dmi_dtx_batched_list)) {
+				dbca = d_list_entry(
+					dmi->dmi_dtx_batched_list.next,
+					struct dtx_batched_commit_args,
+					dbca_link);
+				d_list_del(&dbca->dbca_link);
+				ds_cont_hdl_put(dbca->dbca_cont_hdl);
+				D_FREE_PTR(dbca);
+			}
+
+			break;
+		}
+
+		ABT_thread_yield();
```

Comment:
it will cause this ULT always consume 100% CPU cycle and compete with other ULTs in the pool, better to make it to be periodically scheduled. But I did not see such mechanism provided by ABT, the dss_sleep() cannot do it as well.
not a problem for now, we may see how to improve it later (best way seems  adding a timer to ABT).

Reply:
The dtx_batched_commit() ULT is in charge of the task of DTXs batched commit for the all the containers reside on the given VOS. The ideal model is not time-driven, instead, it should be event (DTX count exceeds the threshold or timer expired) driven. Unfortunately, ABT is busy-pull based model, there is no block-wakeup mechanism in current ABT model, even if the ABT_future_wait() is also busy-wait. I hope that there will be some enhancement for ABT in the future for that, then I will adjust DTX related logic.

Reply:
just checked ABT code, the ABT_future_wait() should be not busy-wait. When calling ABT_future_wait() and the future not set, the caller ULT will be set as ABTI_THREAD_REQ_BLOCK and will not add back to pool, it will be added to the future's wait list and will be wakeup (add back to pool to reschedule) when the future is set.

"event (DTX count exceeds the threshold or timer expired) driven" seems fine, may refine to that later. current ABT mechanism seems enough?

Reply:
Double checked the ABT_future_wait(), you seems right. I will consider to replace ABT_thread_yield() with ABT_future_wait() in my patch if possible.